---
title: 'Denecke et. al 2017: R markdown'
output:
  prettydoc::html_pretty:
    theme: architect
    highlight: github
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=F,message=F)
dir.create("~/test_DGRP")
knitr::opts_knit$set(root.dir = "~/test_DGRP")
```

# Prior To Beginning

Welcome to the R markdown walkthrough for the paper XXXXXXXXXXXXXXXXXXXXXXX (Denecke et. al 2017). This document is designed to guide the reader through every step of the analysis accomplished in the paper and allow for the reproducibility of the study.

Before beginning a few housekeeping issues must be addressed. Firslty is that all code demonstrated here was optimized on a linux machine but has been tested on a mac and a PC. While it should work the particular specifications of the environemnt in which it was written are listed below.
```{r}
version
```


Next the analysis will require several packages. These packages can be installed and loaded into the R enviroment with the following chunk of code.

```{r, Load Packages}

list.of.packages <- c("ggplot2","dplyr","ggplot2","tidyr","qqman","lme4","knitr","RCurl","R.utils")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
library(data.table)
library(tidyr)
library(dplyr)
library(ggplot2)
library(lme4)
library(qqman)
library(knitr)
library(RCurl)
library(R.utils)
```


Finally, the data will have to be downloaded and unzipped in the home directory. Also, during the analysis there are several custom functions that made the writing of the analysis far easier. These functions can all be found in the "functions.list.R" file which we must run before proceeding with the analysis. Finally, the code used for graphing aestitics used in all figures can also be bulky and therefore make the communication of results difficult. I have therefore encoded all ggtheme parameters in a separate file called "themes.R".

```{r Download Data and Import into R}

## Source R codes for plot themes and a few custom functions
themes <- getURL("https://raw.githubusercontent.com/shanedenecke/Rmarkdown_DGRP_Test/master/themes.R")
eval(parse(text = themes))
functions <- getURL("https://raw.githubusercontent.com/shanedenecke/Rmarkdown_DGRP_Test/master/functions.list.R")
eval(parse(text = functions))

# Load the raw data from github
download.file("https://raw.githubusercontent.com/shanedenecke/Rmarkdown_DGRP_Test/master/Denecke_et.al_2017_Raw_Data_Binary",destfile="~/test_DGRP/raw.data")
load("~/test_DGRP/raw.data")


## Import external Mod encode RNAseq dataset if not already present
if(!("mod_encode.tsv" %in% list.files())){
download.file("http://flybase.org/static_pages/downloads/FB2013_03/genes/gene_rpkm_report_fb_2013_03.tsv.gz",destfile="Mod_Encode_Raw.gz")
gunzip(filename="Mod_Encode_Raw.gz",destname="mod_encode.tsv",overwrite=F)
}
mod.encode.raw <- fread("mod_encode.tsv",sep="\t",header=T,showProgress = F)


## Download and read transcriptomic data if not already present
if(!("dgrp.array.exp.female.txt" %in% list.files())){
      download.file("http://dgrp2.gnets.ncsu.edu/data/website/dgrp.array.exp.female.txt",
                    destfile="~/test_DGRP/female.transcriptome.txt",quiet=T)
      download.file("http://dgrp2.gnets.ncsu.edu/data/website/dgrp.array.exp.male.txt",
                    destfile="~/test_DGRP/male.transcriptome.txt",quiet=T)
}

dgrp.female.transcriptome <- fread("female.transcriptome.txt",header=T,showProgress=F)
dgrp.male.transcriptome <-  fread("male.transcriptome.txt",header=T,showProgress=F)


```

The analysis can now begin in ernest. Each section of this document will walk through the analysis of 1 of the 6 results sections presented in the paper. Through the document all Figures and Tables presented in the paper will be generated in what approaches their final, publishable form.

# Section 3.1

Section 3.1 is focussed on looking at the non-genetic characteristics of the distribution of DGRP genotypes. The first thing that must be accomplished is to summarize the wiggle data.  After importing the data, it can be made interpretable using the wiggle.parse() and wiggle.rmr() custom functions. It can then be summarized using tthe dplyr package. We store this result in sum.wiggle.60 containing the mean RMR (and confidence intervals) for each genotype at each dose. Rember RMRs represent:
$$RMR=\frac{\text{Quantity of Movement Without Imidacloprid}}{\text{Quantity of Movement After 1 Hour of Imidacloprid Exposure}}$$ 


```{r Clean and Summarize Wiggle Data}
sum.wiggle.60 <- dgrp.wiggle.raw %>% wiggle.parse(write=F) %>% 
      wiggle.rmr(write=F) %>% filter(time==60) %>%  group_by(genotype,dose) %>% 
      summarize(rmr.ci=conf.diff(rmr),rmr=mean(rmr),raw=mean(raw)) %>% 
      arrange (desc(dose),rmr)
```

Next some modifications will need to be made for plotting Figure 3.1, which reflects the distribution of the DGRP lines. In particular a data.frame will be created called "annot" which will be used to map aestetics to individual facets of the ggplot graph. This will be a common practice throughout the analysis.
```{r Clean Figure 1 Plotting Data}

# Add some modifications that will help for plotting
sum.wiggle.60$rank <- factor(as.character(rep(1:(nrow(sum.wiggle.60)/2),2)),
                             levels=unique(rep(1:(nrow(sum.wiggle.60)/2),2)))
sum.wiggle.60$dose <- factor(sum.wiggle.60$dose,levels=c("25ppm","100ppm"))
sum.wiggle.60$colour <- rep(colorRampPalette(c("steelblue","orangered"))(nrow(sum.wiggle.60)/2),2)
annot <- data.frame(x = rep(-Inf, length(unique(sum.wiggle.60$dose))), 
                    y = rep(Inf, length(unique(sum.wiggle.60$dose))), 
                    dose=as.character(unique(sum.wiggle.60$dose)),
                    labs=LETTERS[1:length(unique(sum.wiggle.60$dose))])

```

Now Figure 1 can be plotted. This figure shows each DGRP genotype (in order) on the x axis and the mean RMR on the y axis. Remember the higher the RMR the more resistant the DGRP line. At both doses there is a range of responses from the super susceptible (blue) to the super resistant (orangered)

```{r Figure 1,fig.width=12,fig.height=5,fig.cap="This is figure 1"}

gp <- ggplot(data=sum.wiggle.60,aes(x=rank,y=rmr,fill=rank),stat="identity")
gp <- gp+facet_grid(~dose)
gp <- gp+geom_bar(position="dodge",stat="identity",colour="black",size=.1)
gp <- gp+geom_errorbar(aes(ymin=rmr+rmr.ci,ymax=rmr-rmr.ci),width=.8,size=.3)
gp <- gp+scale_fill_manual(values=sum.wiggle.60$colour)
gp <- gp+scale_y_continuous(expand = c(0, 0),limits=c(0,1.3),
                            breaks=c(seq(0,1.2,.2)),labels=c(seq(0,1.2,.2)))
gp <- gp+labs(y="Mean Relative Movement Ratio\n",x="\nOrdered Genotype")
gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"), 
                   inherit.aes=FALSE,vjust=1.2,hjust=-.2)
gp <- gp+theme_bw(base_size=14,base_family="serif")
gp <- gp+spread_theme

print(gp)
```

Before moving on to trying to understanding the genetics underlying these differences, some basic checks must be performed on the data. One such check is to see how related our two doses are. This is shown in figure S1 by creating a scatterplot showing the RMRs at 25ppm vrs the RMRs at 100ppm. 

```{r Figure S1}
sum.wiggle.60 <- arrange(sum.wiggle.60,genotype)
dose.compare <- data.frame(genotype=unique(sum.wiggle.60$genotype),
                           high.dose=filter(sum.wiggle.60,dose=="100ppm")$rmr,
                           low.dose=filter(sum.wiggle.60,dose=="25ppm")$rmr)
gp <- ggplot(dose.compare,aes(x=high.dose,y=low.dose))
gp <- gp+geom_point(size=2)
gp <- gp+geom_smooth(method="lm")
gp <- gp+labs(y="100ppm Relative Movement Ratio\n",x="\n25ppm Relative Movement Ratio")
gp <- gp+theme_bw(base_size=14,base_family="serif")
gp <- gp+corr_theme
print(gp) 
```

There appears to be quite a strong correlation. Another point of interest would be to see if there is any correlation between the initial motility (in the absence of insecticides) and the final RMR. As the Wiggle Index is relatively new, we must be sure that our RMR is not an artifact of some lines moving more than others in general. Figure S2 shows the initial motility of each genotype on the x axis and the final RMR (imidacloprid response) on the y axis.

```{r Figure S2}
sum.wiggle <- dgrp.wiggle.raw %>% wiggle.parse() %>% wiggle.rmr() %>% group_by(genotype,time,dose) %>% 
      summarize(rmr=mean(rmr),raw=mean(raw)) %>% arrange (desc(time),genotype)

initial.compare <- sum.wiggle %>% gather(mes,output,rmr,raw) %>% filter((time=="0" & mes=="raw") | (time=="60" & mes=="rmr")) 
initial.compare <- data.frame(genotype=initial.compare$genotype,initial.motility=filter(initial.compare,mes=="raw")$output,wiggle.response=filter(initial.compare,mes=="rmr")$output)


gp <- ggplot(initial.compare,aes(x=initial.motility,y=wiggle.response))
gp <- gp+geom_point(size=2)
gp <- gp+geom_smooth(method="lm")
gp <- gp+ylab("Wiggle Response\n")
gp <- gp+xlab("\nInitial Motility")
gp <- gp+theme_bw(base_size=14,base_family="serif")
gp <- gp+corr_theme
print(gp) 

```

There doesn't seem to be much of an effect. This means that the estimate of imidacloprid resistance is not affected by the baseline differences in motility between genotypes.

Next heritability can be estimated. This compares the variance between genotypes to variance within genotypes. By doing this the heritability can be measured.

```{r Calculate Heritability}
corrected.wiggle.data <- dgrp.wiggle.raw %>% wiggle.parse() %>% wiggle.rmr()  %>% filter(time==60)
heritability.list=list()
doses <- c("25ppm","100ppm")
for (a in doses){
      heritability.list[[a]]=list()
      sub.dose <- filter(corrected.wiggle.data,dose==a)
      sub.rmr <- sub.dose$rmr
      test <- lmer(sub.rmr~1+(1|sub.dose$genotype))
      heritability.list[[a]] <- c(genetic.variance=NA,environmental.variance=NA,heritability=NA)
      heritability.list[[a]]["genetic.variance"] <- VarCorr(test)$'sub.dose$genotype'[1]
      heritability.list[[a]]["environmental.variance"] <- heritability.list[[a]]["genetic.variance"]+
            attr(VarCorr(test),"sc")^2
      heritability.list[[a]]["heritability"] <- heritability.list[[a]]['environmental.variance']/
            (heritability.list[[a]]["environmental.variance"]+heritability.list[[a]]["genetic.variance"])
} 

heritability.list <- data.frame(heritability.list)
colnames(heritability.list) <- c("25ppm","100ppm")
kable(heritability.list,align="c")
```

The final part of this section will clean the data in a format that can be read by the GWAS pipeline. The mean RMRs for each genotype was submitted separately to http://dgrp2.gnets.ncsu.edu/ for further processing.The first few rows of the submission files are printed below. These are tables S2 and S3

```{r Clean phenotype data to submit}
for(d in unique(sum.wiggle.60$dose)){
      single.dose <- sum.wiggle.60 %>% filter(dose==d) %>% select(genotype,rmr)
      single.dose[['genotype']] <-gsub("RAL_","",single.dose[['genotype']]) 
      clean.matrix <- matrix(as.numeric(as.matrix(single.dose)),ncol=2)
      print(head(clean.matrix))
}             
```


#Section 3.4

In section 3.4 a subset of the DGRP lines were used to assess only imidacloprid metabolism in the DGRP. For 18 genotypes (9 resistant and 9 susceptible) amounts of imidacloprid, IMI-5-OH and IMI-Olefin were mesured in both media and larval bodies. 

Before plotting the raw data needs cleaning. The code chunk below accomplishes this nicely.

```{r Clean DGRP HPLC Data}

dgrp.hplc.data.rearrange <- dgrp.hplc.data %>% 
      gather("mes.type","mes.value",larvae,media) %>%
      filter(!is.na(mes.value)) %>%
      arrange(rank) 

dgrp.hplc.summary <- dgrp.hplc.data.rearrange %>% 
      group_by(genotype,metabolite,mes.type) %>%
      summarize(mes.value=mean(mes.value)) 
      wiggle.hplc.subset <- sum.wiggle.60 %>% 
      filter(dose=="25ppm") %>% 
      filter(genotype %in% 
      unique(dgrp.hplc.summary$genotype)) %>% select(genotype,rmr)

dgrp.hplc.summary <- full_join(wiggle.hplc.subset,dgrp.hplc.summary,by="genotype")

dgrp.hplc.plot.data <- dgrp.hplc.data[-intersect(which(dgrp.hplc.data$metabolite=="Imidacloprid"),which(dgrp.hplc.data$mes.type=="media")),]
dgrp.hplc.plot.data$genotype <- factor(dgrp.hplc.plot.data$genotype,levels=unique(dgrp.hplc.plot.data$genotype))
dgrp.hplc.plot.data$metabolite <- factor(dgrp.hplc.plot.data$metabolite,levels=c("Imidacloprid","5-OH","Olefin"))
dgrp.hplc.plot.data$mes.type <- factor(dgrp.hplc.plot.data$mes.type)

```

Some further variables are needed for plotting Figure 3. As was done prevously the annot data frame will map data to each facet. In this case the significance of the linear model predicting the level of a chemical with RMR at 25ppm is represented by signficance stars. The following code accomplishes this task.

```{r Plot HPLC Data}

vars <- data.frame(expand.grid(levels(dgrp.hplc.plot.data$metabolite), levels(dgrp.hplc.plot.data$mes.type)))
colnames(vars) <- c("metabolite", "mes.type")

annot <- data.frame(x = rep(Inf, dim(vars)[1]), 
                    y = rep(Inf, dim(vars)[1]), 
                    vars,
                    labs=LETTERS[1:dim(vars)[1]])
equation.list <- vector("character",6)
for(m in unique(dgrp.hplc.plot.data$metabolite)){
      for(t in (unique(dgrp.hplc.plot.data$mes.type))){
            sub <- subset(dgrp.hplc.summary,metabolite==m & mes.type==t)
            equation.list[intersect(which(annot$metabolite==m),which(annot$mes.type==t))] <- equation.extract.hplc("mes.value","rmr",data=sub,sub=F)
      }}
equation.list[intersect(which(annot$metabolite=="Imidacloprid"),which(annot$mes.type=="media"))] <- ""
annot <- mutate(annot,equation.list=equation.list)
dot.col <- c(rep("steelblue2",180),rep("orangered2",175))
```

Now Figure 3 can be plotted. This figure will be in the same format as Figure 5 and Figure S5. Each will have 6 panels with 2 columns representing measurments in the larval body or media and 3 rows reflecting the 3 chemicals measured. 

```{r Plot Figure 3,fig.height=8,fig.width=10}
gp <- ggplot(dgrp.hplc.plot.data,aes(x=genotype,y=mes.value)) 
      gp <- gp+facet_grid(metabolite~mes.type,scales="free_y")
      gp <- gp+geom_dotplot(binaxis='y', stackdir='center',colour="black",dotsize=1.2,fill=dot.col) 
      gp <- gp+labs(y="Quantity of Chemical",x="Genotype")
      gp <- gp+annotate("rect",xmin=-Inf,xmax=9.5,ymin=-Inf,ymax=Inf,fill="steelblue2",alpha=.3)
      gp <- gp+annotate("rect",xmin=9.5,xmax=Inf,ymin=-Inf,ymax=Inf,fill="orangered2",alpha=.3)
      gp <- gp+stat_summary(geom="errorbar",colour='black',size=.75,width=.6,fun.data="mean_se")
      gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.6,size=.3)
      gp <- gp+geom_dotplot(binaxis='y', stackdir='center',colour="black",dotsize=1.2,fill=dot.col) 
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),vjust=1.2,hjust=1.2)
      gp <- gp+geom_text(size=7,data=annot,aes(9.5, y, label=equation.list,fontface="bold",family="serif"),vjust=1.2)
      gp <- gp+theme_bw(base_size=14,base_family="serif")
      gp <- gp+dgrp.hplc_theme
print(gp)
```


#Section 3.2

This section will cover section 3.2 in the paper which deals with the genome wide association study. First all the appropriate data must be imported into R. This includes two tables showing all assocaitions in the GWAS, two tables showing only the significant associations in the GWAS, and modEncode expression data obtained from XXXXXXXXXXXx.

One way to vizualize the results of a GWAS is to plot the association of each variant with the phenotype. Before doing this the data must be cleaned and annotated in a format that can be plotted.

```{r Clean GWAS assocaition data}
assoc.list <- vector('list',2)
names(assoc.list) <- c("25ppm","100ppm")

for (d in doses){  
      all.associations <- get(paste("gwas.all.associations.",d,sep=""))
      all.associations <- select(all.associations,ID,SinglePval)
      variant <- unlist(strsplit(all.associations$ID,split="_"))
      chromosome <- variant[seq(1, length(variant), 3)]
      base <- as.numeric(variant[seq(2, length(variant), 3)])
      chromosome <- multi.gsub(chromosome,find=c("X",'2L','2R','3L','3R'),replace=c(1,2,3,4,5))
      gwas.plot.data <- all.associations %>% 
            mutate(CHR=as.numeric(chromosome),BP=base) %>% 
            select(SinglePval,ID,CHR,BP)
      colnames(gwas.plot.data)=c('P','SNP','CHR','BP')
      variant.sig <- gwas.plot.data$SNP[which(gwas.plot.data$P<=.05/length(gwas.plot.data$SNP))]
      #gwas.plot.data <- filter(gwas.plot.data,P<.0001)
      assoc.list[[d]] <- gwas.plot.data
}
```

Now the data can be vizualized. First, a Manhattan plot can becreated which shows every variant in order according to its chromosomal position along the x axis. The y axis displays the significance of each association.

```{r QQ Plots,fig.width=10,cache=T}

par(mar=c(5,6,2,1),mgp=c(3,1,0),mfrow=c(1,2),bty="l") 

qq(assoc.list[["25ppm"]]$P, main="A", family="serif",cex.main=2,font.axis=2,cex.lab=1.5,cex.axis=1.5,
      ylab=list(expression(bold(paste("Observed -log"[10]*"(p)"))),font=2),
      xlab=list(expression(bold(paste("Expected -log"[10]*"(p)"))),font=2))

qq(assoc.list[["100ppm"]]$P,main="B", family="serif",cex.main=2,font.axis=2,cex.lab=1.5,cex.axis=1.5,
   ylab=list(expression(bold(paste("Observed -log"[10]*"(p)"))),font=2), 
   xlab=list(expression(bold(paste("Expected -log"[10]*"(p)"))),font=2))  

```

```{r Manhattan Plots,fig.width=10,cache=T}

par(mar=c(5,6,2,1),mgp=c(3,1,0),mfrow=c(1,2))
manhattan(assoc.list[["25ppm"]],col=c("black","orange"),
          genomewideline=-log10(.05/length(assoc.list[["25ppm"]]$SNP)),
          chrlabs=c("X",'2L','2R','3L','3R'),
          highlight=variant.sig,main="A",cex.main=2,ylim=c(0,-log10(1e-8)),
          family="serif",font=2,cex.lab=1.5,cex.axis=1.5,
          xlab="",ylab=list(expression(bold(paste("-log"[10]*"(p)"))),font=2))  

manhattan(assoc.list[["100ppm"]],col=c("black","orange"),
          genomewideline=-log10(.05/length(assoc.list[["100ppm"]]$SNP)),
          chrlabs=c("X",'2L','2R','3L','3R'),
          highlight=variant.sig,main="B",cex.main=2,ylim=c(0,-log10(1e-8)),
          family="serif",font=2,cex.lab=1.5,cex.axis=1.5,
          xlab="",ylab=list(expression(bold(paste("-log"[10]*"(p)"))),font=2))
```


These plots show that some variatns are significantly associated with imidacloprid resistance. Which variants are these? By narrowing the search to anything below p= 10^-5 the most significant variants can be associated along with their nearest gene.

```{r Creat Table 1}

low.dose.gwas <- dgrp.gwas.parse(gwas.top.candidates.25ppm) %>% mutate(dose="25ppm")
high.dose.gwas <- dgrp.gwas.parse(gwas.top.candidates.100ppm) %>% mutate(dose="100ppm")
table.1 <- rbind(low.dose.gwas,high.dose.gwas) %>% select(dose,ID,Gene.Name,MAF,SingleEff,SinglePval)
colnames(table.1) <- c("dose","Variant ID","gene","maf","effect size","p-value")
kable(table.1)


```


The genes in this list are the best estimate of the genes involved in imidacloprid resistance within the DGRP. Because there are no genes known to have roles in resistance, looking at this list as a group is helpful. For instance these genes tend to be upregulated in the third instar larvae central nervous system compared to the genome wide average. This can be assessed using the custom mod.encode.enrich() function.

```{r Mod Encode Expression enrichment of GWAS Candidates,cache=T}
unique.gwas.list <- unique(c(low.dose.gwas$Gene.Name,high.dose.gwas$Gene.Name))
unique.gwas.list <- unique.gwas.list[unique.gwas.list!="intergenic"]
l3.enrichment <- mod.encode.enrich(raw.data.file=mod.encode.raw,
                  gene.list = unique.gwas.list, 
                  parent.library="modENCODE_mRNA-Seq_tissues",
                  stage.search.string="L3",
                  tissue.search.string="CNS")
l3.enrichment <- as.data.frame(t(as.matrix(l3.enrichment)))
rownames(l3.enrichment) <- NULL
kable(l3.enrichment)
```

As the last thing that will be done in section 3.2, a few of the larger objects present in the workspace can be cleared.

```{r}

rm(list=c("mod.encode.raw","gwas.all.associations.25ppm","gwas.all.associations.100ppm","assoc.list"))

```

#Section 3.3

Section 3.3 considers the transcriptomic dataset generated by Huang et. al (2015). The object here is to correlate transcriptomic variation with phenotypic variation, much like was done in the previous section with genomic variation. 


The transcriptomic datasets can be downloaded from http://dgrp2.gnets.ncsu.edu/data.html and are composed of two replicates of each transcriptome for each sex. Because this study considered larval samples that were not sexed, all 4 replicates of the transcriptomes were averaged together with the custom function transcript.combine() and a series of operations connected by the %>% operator. 

```{r Combine and Clean Transcriptomic Data,cache=T}

combined.transcriptomes <- transcript.combine(male=dgrp.male.transcriptome,female=dgrp.female.transcriptome)
combined.transcriptomes <- merge(combined.transcriptomes,fbgid.symbol.key,by="gene")

clean.transcriptome.data <- combined.transcriptomes %>% 
      gather(key="Genotype",value="RPKM",RAL_21:RAL_913) %>%
      group_by(symbol,Genotype) %>% 
      summarize(avg=mean(RPKM,na.rm=T)) %>%
      spread(key=Genotype,value=avg)

```

A few modifications need to be performed on this dataset including combining it with the phenotypic data. 

```{r Subset Transcriptomic and Phenotypic Data}

# Subset the genotypes which are in both the phenotyic and genotypic datasets
transcriptome.genotypes <- colnames(clean.transcriptome.data)[-1]
wiggle.genotypes <- as.character(sum.wiggle.60$genotype)
common.genotypes <- intersect(transcriptome.genotypes,wiggle.genotypes)
phenotype.data.subset <- sum.wiggle.60[wiggle.genotypes %in% common.genotypes,]
transcriptome.subset.symbol <- data.table(data.frame(clean.transcriptome.data)[,c(TRUE,transcriptome.genotypes %in% common.genotypes)])

## Run a linear model predicting RMR with each transcript in the transcriptome
symbol <- transcriptome.subset.symbol$symbol
transcriptome.subset <- select(transcriptome.subset.symbol,-symbol)
transcriptome.test <- sapply(select(transcriptome.subset,starts_with("RAL")),as.numeric)
table_2 <- vector('list',2)
names(table_2) <- doses
```

Once this is done the transcriptome wide association study can be run. This code fits a linear model predicting the mean RMR at either 25 or 100ppm with the magnitude of the expression of a given transcript.


```{r Run TWAS,cache=T}
for (d in doses){
      sub.phen.dose <- as.numeric(subset(phenotype.data.subset,dose==d)$rmr)
      output=matrix(NA,nrow=length(transcriptome.test[,1]),ncol=3)
      colnames(output)=c("estimate","pval","adjrsq")
      for (i in 1:nrow(transcriptome.test)) {
            test <- summary(lm(sub.phen.dose~transcriptome.test[i,]))
            est <- test$coefficients[2,1]
            pval <- test$coefficients[2,4]
            rsq <- test$adj.r.squared
            output[i,] <- c(est,pval,rsq)
      }
      
      rownames(output)=symbol
      twas.summary <- data.frame(output) %>%
            mutate(gene=symbol,dose=d) %>%
            filter(pval<1e-3) %>% arrange(pval) %>%
            select(dose,gene,pval,estimate,adjrsq)
      colnames(twas.summary) <- c("dose","gene","p-value","effect size","adjusted R squared")
      table_2[[d]] <- twas.summary
}
kable(rbindlist(table_2,use.names=T))
```

At each dose Cyp6g1 appears to be the most significnat assocaition. Also present is its genomic neighbor Cyp6g2. To plot these associations individually a few operations can be used to create a data frame that contains just the expression of thse two genes and the phenoyptic Wiggle Index data. As was done before, a data frame "annot" can be genrated which contains information used to plot each facet of the graph. In this case we can display the equation of the linear model used.


```{r Clean and Annotate Cyp6g1 and Cyp6g2 Association Plots}
g1.haplotype.sub <- g1.haplotypes %>% filter(genotype%in%common.genotypes)
individual.transcripts <- transcriptome.subset.symbol %>% filter(symbol=="Cyp6g1"|symbol=="Cyp6g2") %>%
      gather(genotype, val, 2:(length(common.genotypes)+1)) %>%  
      spread(symbol, val) %>% arrange(genotype) %>% group_by(genotype) %>% summarise_each(funs(sum(., na.rm=TRUE))) %>%
      full_join(select(filter(phenotype.data.subset,dose=="100ppm"),genotype,rmr),by="genotype") %>% 
      full_join(g1.haplotype.sub,by="genotype") %>% gather(key="transcript",value="expression",Cyp6g1,Cyp6g2)


individual.transcripts$transcript <- factor(individual.transcripts$transcript,levels=c("Cyp6g1","Cyp6g2"))
equation.list <- vector("character",2)
for (t in unique(individual.transcripts$transcript)) equation.list[which(t==unique(individual.transcripts$transcript))] <- equation.extract("rmr","expression",individual.transcripts,"transcript",t)
annot <- data.frame(x = rep(-Inf, length(unique(individual.transcripts$transcript))), 
                    y = rep(Inf, length(unique(individual.transcripts$transcript))), 
                    transcript=as.character(unique(individual.transcripts$transcript)),
                    equation=equation.list,
                    labs=LETTERS[1:length(unique(individual.transcripts$transcript))])
annot$transcript <- factor(annot$transcript,levels=annot$transcript)

```

Now the associations of both Cyp6g1 and Cyp6g2 can be vizualised according to haplotypes.

```{r Plot Figure 2,fig.width=16,fig.height=6}
gp <- ggplot(individual.transcripts,aes(x=expression,y=rmr))
      gp <- ggplot(individual.transcripts,aes(x=expression,y=rmr))
      gp <- gp+facet_grid(facets=.~transcript,scales="free_x")
      gp <- gp+geom_point(size=4,aes(shape=haplotype,colour=haplotype))
      gp <- gp+scale_shape_identity()
      gp <- gp+geom_smooth(method="lm")
      gp <- gp+labs(y="Relative Movement Ratio\n",x="\nTranscript Expression Expression (log2 RPKM)")
      gp <- gp+scale_colour_manual(values=c("red2","blue","green4","black"))
      gp <- gp+ylim(c(0,1.2))
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"), inherit.aes=FALSE,vjust=1.2,hjust=-.2)
      gp <- gp+geom_text(size=4,data=annot,aes(Inf, Inf, label=equation,family="serif"), inherit.aes=FALSE,vjust=1.2,hjust=1.2)
      gp <- gp+theme_bw(base_size=14,base_family="serif")
      gp <- gp+cyp6g1.ind_theme
print(gp) 
```

Lastly, as was done for section 3.2, some of the larger objects can be removed from the R environment. 

```{r Remove bulky transcriptome datasets from global environemnt}
rm(list=c("dgrp.male.transcriptome","dgrp.female.transcriptome"))
```

#Section 3.5

From section 3.3 it was shown that Cyp6g1 is the transcript most significantly associated with imidacloprid resistance. Removing Cyp6g1 then follows on as a logical next step.


First the Wiggle data can be considered. The usual cleaning steps can be applied to this dataset as well.

```{r Clean Cyp6g1KO Wiggle Data}
cyp6g1.ko.wiggle.plot <- cyp6g1.ko.wiggle %>% filter(time==60) %>% select(genotype,dose,time,rmr,identifier)
cyp6g1.ko.wiggle.plot$genotype <- factor(cyp6g1.ko.wiggle.plot$genotype,
                                         levels=c("RAL_517","RAL_517-Cyp6g1KO","Canton-S","Canton-S-Cyp6g1KO","Wxac","Wxac-Cyp6g1KO"))
cyp6g1.ko.wiggle.plot$identifier <- factor(cyp6g1.ko.wiggle.plot$identifier,
                                           levels=c("RAL_517 (25ppm)","RAL_517 (100ppm)","Canton-S (5ppm)","Wxac (5ppm)"))
vars <- data.frame(expand.grid(levels(cyp6g1.ko.wiggle.plot$identifier)))
colnames(vars) <- c("identifier")
annot <- data.frame(x = rep(Inf, dim(vars)[1]), y = rep(Inf, dim(vars)[1]), vars,labs=LETTERS[1:dim(vars)[1]])

equation.list <- vector("character",4)
for(i in unique(levels(cyp6g1.ko.wiggle.plot$identifier))){
            sub <- subset(cyp6g1.ko.wiggle.plot,identifier==i)
            equation.list[which(annot$identifier==i)] <- equation.extract.hplc("rmr","genotype",data=sub,sub=F,bonf=4)
      }
annot$pval <- equation.list

```

Now Figure 4 can be plotted. This figure will show the wiggle response of of Cyp6g1KOs and their controls in 3 different backgrounds. Panels A and B show two doses of RAL517-Cyp6g1KO and C and D show Canton-S-Cyp6g1KO and Wxac-Cyp6g1KO respectively


```{r Plot Figure 4,fig.width=10}
gp <- ggplot(cyp6g1.ko.wiggle.plot,aes(x=genotype,y=rmr,fill=genotype))
      gp <- ggplot(cyp6g1.ko.wiggle.plot,aes(x=genotype,y=rmr,fill=genotype))
      gp <- gp+facet_grid(.~identifier,scales="free_x")
      gp <- gp+stat_summary(geom="errorbar",colour='black',size=1,width=.3,fun.data="mean_cl_normal")
      gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.4,colour='black',size=.8)
      gp <- gp+geom_dotplot(binaxis='y',stackdir="center",dotsize=1.2)
      gp <- gp+scale_fill_manual(values=c("blue4","blue1","purple4","purple1","gold4","gold1")) 
      gp <- gp+labs(y="Relative Movement Ratio\n",x="\nGenotype")
      gp <- gp+ylim(c(0,1.4))
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),inherit.aes=F,vjust=1.2,hjust=1.2)
      gp <- gp+geom_text(size=7,data=annot,aes(1.5, y, label=pval,fontface="bold",family="serif"),inherit.aes=F,vjust=1.2)
      gp <- gp+theme_bw(base_size=14,base_family="serif")
      gp <- gp+g1KO.wiggle_theme
print(gp)
```

Since RAL517-Cyp6g1KO was the only one to show any differnece in Wiggle Index response, only this knockout was considered for HPLC-MS analysis. First the data must be cleaned

```{r Clean Figure 5 Data}

cyp6g1KO.hplc.data.plot <- cyp6g1KO.hplc.data[-intersect(which(cyp6g1KO.hplc.data$metabolite=="Imidacloprid"),which(cyp6g1KO.hplc.data$measurement_type=="Media")),]
cyp6g1KO.hplc.data.plot$genotype <- factor(cyp6g1KO.hplc.data.plot$genotype,levels=c("RAL_517","RAL_517-Cyp6g1KO"))
cyp6g1KO.hplc.data.plot$metabolite <- factor(cyp6g1KO.hplc.data.plot$metabolite,levels=c("Imidacloprid","5-OH","Olefin"))
vars <- data.frame(expand.grid(levels(cyp6g1KO.hplc.data.plot$metabolite), levels(as.factor(cyp6g1KO.hplc.data.plot$measurement_type))))
colnames(vars) <- c("metabolite", "measurement_type")
annot <- data.frame(x = rep(-Inf, dim(vars)[1]), y = rep(Inf, dim(vars)[1]), vars,labs=LETTERS[1:dim(vars)[1]])

equation.list <- vector("character",6)
for(m in unique(cyp6g1KO.hplc.data$metabolite)){
      for(t in (unique(cyp6g1KO.hplc.data$measurement_type))){
            sub <- subset(cyp6g1KO.hplc.data,metabolite==m & measurement_type==t)
            equation.list[intersect(which(annot$metabolite==m),which(annot$measurement_type==t))] <- equation.extract.hplc("measurement_value","genotype",data=sub,sub=F,bonf=6)
      }}
equation.list[intersect(which(annot$metabolite=="Imidacloprid"),which(annot$mes.type=="media"))] <- ""
annot <- mutate(annot,pval=equation.list)
```

Much like Figure 3 the amount of each imidacloprid derivative can be vizualized for each genotype in the larval body and excreta. 

```{r Plot Figure 5,fig.height=8,fig.width=5}
  gp <- ggplot(cyp6g1KO.hplc.data.plot,aes(x=genotype,y=measurement_value,fill=genotype))
      gp <- ggplot(cyp6g1KO.hplc.data.plot,aes(x=genotype,y=measurement_value,fill=genotype))
      gp <- gp+facet_grid(metabolite~measurement_type,scales="free_y")
      gp <- gp+labs(y="Quantity of Chemical\n",x="\nGenotype")
      gp <- gp+stat_summary(geom="errorbar",colour='black',size=1,width=.4,fun.data="mean_se")
      gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.4,size=.3)
      gp <- gp+geom_dotplot(binaxis='y', stackdir='center',dotsize=1.5) 
      gp <- gp+scale_fill_manual(values=c("blue4","blue1"))
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),inherit.aes=FALSE,vjust=1.2,hjust=-.2)
      gp <- gp+geom_text(size=7,data=annot,aes(1.5, y, label=pval,fontface="bold",family="serif"),inherit.aes=F,vjust=1.2)
      gp <- gp+theme_bw(base_size=14,base_family="serif")
      gp <- gp+g1KO.wiggle_theme
print(gp)
```

#Section 3.6

In section 3.6 the effect of other P450s on imidacloprid resistance is examined. in particular the genomic neighbors of Cyp6g1, Cyp6g2 and Cyp6t3.

The expression of Cyp6g1, Cyp6g2 and Cyp6t3 was measured in the digestive tissues (midguts and tubules) of 4 DGRP lines, two with M haplotypes and two with AA haplotypes. The custom function rt.data.parse takes raw output from XXXXX and converts it into a useable format. Biologcial replicates 3 and 4 from RAL_360 were removed due to the high variability in the housekeeper gene RP49.

```{r Parse Real Time Data}
parsed.rt.data <- rt.data.parse(data=P450.real.time.data)
filtered.data <- parsed.rt.data[which(!((parsed.rt.data$biological.rep=="3" | parsed.rt.data$biological.rep=="4") & parsed.rt.data$Sample=="360")),]

```

Next the real time PcR data can be processed in parallel with expression data from section 3.3. Cleaning this data and doing some basic conversions can reveal what the fold upregulation of each gene is compared to a line that expresses low levels of each (RAL_776).

```{r Summarize Expression DAta}

sum.rt.data <- filtered.data %>% 
      group_by(Target,Sample,biological.rep) %>% 
      summarize(mean.ct=mean(Cq)) %>% 
      spread(key=Target,value=mean.ct) %>% 
      mutate(g1dct=-(RP49-CYP6G1),g2dct=-(RP49-CYP6G2),t3dct=-(RP49-CYP6T3)) %>% 
      mutate(genotype=as.character(paste("RAL_",Sample,sep=""))) %>%
      mutate(haplotype=allele.ad(genotype)) %>% data.table() %>%
      mutate(g1ddct=g1dct-mean(.[which(.[["genotype"]]=="RAL_776"),][["g1dct"]],na.rm=T)) %>% 
      mutate(g2ddct=g2dct-mean(.[which(.[["genotype"]]=="RAL_776"),][["g2dct"]],na.rm=T)) %>% 
      mutate(t3ddct=t3dct-mean(.[which(.[["genotype"]]=="RAL_776"),][["t3dct"]],na.rm=T)) %>% 
      mutate(Cyp6g1=2^-g1ddct,Cyp6g2=2^-g2ddct,Cyp6t3=2^-t3ddct) %>% 
      mutate(type="Digestive Tissues Larvae") %>%
      select(genotype,haplotype,type,Cyp6g1:Cyp6t3) %>%
      gather(key="gene",value="relative.expression",Cyp6g1:Cyp6t3)


dgrp.p450.trans <- combined.transcriptomes %>%
      select(symbol,RAL_138,RAL_360,RAL_776,RAL_843) %>%
      filter(symbol %in% c("Cyp6g1","Cyp6g2","Cyp6t3")) %>% 
      transpose.collapse(column="symbol") %>%
      mutate(haplotype=allele.ad(genotype)) %>%
      mutate(Cyp6g1=2^(Cyp6g1-mean(.[which(.[["genotype"]]=="RAL_776"),][["Cyp6g1"]],na.rm=T))) %>%
      mutate(Cyp6g2=2^(Cyp6g2-mean(.[which(.[["genotype"]]=="RAL_776"),][["Cyp6g2"]],na.rm=T))) %>%
      mutate(Cyp6t3=2^(Cyp6t3-mean(.[which(.[["genotype"]]=="RAL_776"),][["Cyp6t3"]],na.rm=T))) %>%
      mutate(type="Whole Body Adults") %>%
      select(genotype,haplotype,type,Cyp6g1:Cyp6t3) %>%
      gather(key="gene",value="relative.expression",Cyp6g1:Cyp6t3) 

```

Next a few operations can be applied to prepare the data for plotting

```{r Clean Expression Data and Prepare for Plotting}

combined.data <- rbindlist(list(sum.rt.data,dgrp.p450.trans),use.names=T)
combined.data <- data.frame(combined.data)[complete.cases(combined.data),]
combined.data$genotype <- factor(combined.data$genotype,levels=c("RAL_776","RAL_843","RAL_138","RAL_360"))
combined.data$haplotype <- factor(combined.data$haplotype,levels=c("M","AA"))

vars <- data.frame(expand.grid(unique(combined.data$gene), unique(combined.data$type)))
colnames(vars) <- c("gene", "type")
annot <- data.frame(x = rep(-Inf, dim(vars)[1]), y = rep(Inf, dim(vars)[1]), vars,labs=LETTERS[1:dim(vars)[1]])

```

Finally, Figure 6 can be plotted. This figure will reflect the fold increase of each measurment compared to the mean of the measurment in RAL_776. Each gene will be considered in a separate panel along with whether the measuremnts were made by real time PCR in the digestive tissues (measured in this study) or weather it was obtained from Huang et. al (2015) and reflects whole body adult transcriptomes.

```{r Plot Figure 6,fig.height=6,fig.width=12}
gp <- ggplot(data=combined.data,aes(x=genotype,y=relative.expression,fill=haplotype))
      gp <- gp+facet_grid(type~gene,scales="free_y")
      gp <- gp+geom_dotplot(binaxis='y',stackdir="center",dotsize=1.5)
      gp <- gp+annotate("rect",xmin=-Inf,xmax=2.5,ymin=-Inf,ymax=Inf,fill="steelblue2",alpha=.3)
      gp <- gp+annotate("rect",xmin=2.5,xmax=Inf,ymin=-Inf,ymax=Inf,fill="orangered2",alpha=.3)
      gp <- gp+stat_summary(geom="errorbar",colour='black',size=1,width=.4,fun.data="mean_se")
      gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.4,size=.3)
      gp <- gp+geom_dotplot(binaxis='y',stackdir="center",dotsize=1.5)
      gp <- gp+scale_fill_manual(values=c("steelblue","orangered"))
      gp <- gp+labs(y="Expression Relative to RAL_776\n",x="\nGenotype")
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),inherit.aes=F,vjust=1.2,hjust=-.2)
      gp <- gp+rt_theme
print(gp)
```

The differnece between genotypes for figure 6 can be best estimated by using analysis of variance followed by Tukeys honestly significnat difference. The value in each column represents the p-value relfected by each comparison.

```{r}

TableS4 <- list()
for(t in unique(combined.data$type)){
      for (g in unique(combined.data$gene)){
            sub <- subset(combined.data,type==t & gene==g)
            mod <- with(sub,aov(relative.expression~genotype))
            TableS4[[paste(t,g)]] <- TukeyHSD(x=mod, which='genotype', conf.level=0.95)$genotype[,"p adj"]
      }}
TableS4 <- data.frame(TableS4)
colnames(TableS4) <- c("Dig.Tis.6g1","Dig.Tis.6g2","Dig.Tis.t3","Body.6g1","Body.6g2","Body.t3")
kable(TableS4)
```

Moving on from expression data, the capacities of Cyp6g1 and Cyp6g2 to confer resistance can be directly measured by expressing each genes from a common insertion site using GAL4-UAS. 

First the data must be clenaed

```{r Clean Up Figure 7}

uas.wiggle <- uas.cyp6g12.wiggle %>% wiggle.parse(write=F) %>% wiggle.rmr(write=F) %>% filter(time==60)
uas.wiggle$genotype <- multi.gsub(uas.wiggle$genotype,c("x"),c(" x "))
uas.wiggle$genotype <- factor(uas.wiggle$genotype,levels=c("HRG4 x UAS-NULL","HRG4 x UAS-Cyp6g1","HRG4 x UAS-Cyp6g2"))

annot <- data.frame(x = rep(-Inf, length(unique(uas.wiggle$dose))), 
                    y = rep(Inf, length(unique(uas.wiggle$dose))), 
                    dose=as.character(unique(uas.wiggle$dose)),
                    labs=LETTERS[1:length(unique(uas.wiggle$dose))])

```

Then the data can be plotted considering two discriminatory doses (20 and 40ppm) separatly.

```{r Plot Figure 7}
  gp <- ggplot(uas.wiggle,aes(x=genotype,y=rmr,fill=genotype))
      gp <- gp+facet_grid(.~dose,scales="fixed")
      gp <- gp+stat_summary(geom="errorbar",colour='black',size=1,width=.3,fun.data="mean_cl_normal")
      gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.4,colour='black',size=.8)
      gp <- gp+geom_dotplot(binaxis='y',stackdir="center",dotsize=1.2)
      gp <- gp+scale_fill_manual(values=c("black","magenta1","orange1")) 
      gp <- gp+labs(y="Relative Movement Ratio\n",x="\nGenotype")
      gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),inherit.aes=FALSE,vjust=1.2,hjust=-.2)
      gp <- gp+theme_bw(base_size=14,base_family="serif")
      gp <- gp+g1KO.wiggle_theme
print(gp)
      
```

As was done for Table S4 Tukeys post hoc HSD can be used to observe the significant differences (or lack thereof) between the three genoytpes.

```{r}
TableS5 <- list()
for (d in unique(uas.wiggle$dose)){
      sub <- subset(uas.wiggle,dose==d)
      mod <- with(sub,aov(rmr~genotype))
      TableS5[[d]] <- TukeyHSD(x=mod, which='genotype', conf.level=0.95)$genotype[,"p adj"] 
      
}
TableS5 <- data.frame(TableS5)
colnames(TableS5) <- c("20ppm","40ppm")
kable(TableS5)
```

It appears that Cyp6g2 can confer resistance to imidacloprid. To see whether this observation is mediated by metabolism HPLC-MS can be applied to a genoytpe expressing Cyp6g2. 

First the data must be cleaned.

```{r Clean up Figure S5 Data}

uas.cyp6g2.hplc.data.rearrange <- gather(uas.cyp6g2.hplc.data,key="measurement_type",
                               value="measurement_value",larvae,media)
plot.data <- uas.cyp6g2.hplc.data.rearrange[complete.cases(uas.cyp6g2.hplc.data.rearrange),]
plot.data <- uas.cyp6g2.hplc.data.rearrange[-intersect(which(uas.cyp6g2.hplc.data.rearrange$metabolite=="Imidacloprid"),
                                             which(uas.cyp6g2.hplc.data.rearrange$measurement_type=="media")),]
plot.data$genotype <- factor(plot.data$genotype,levels=c("w1118","Cyp6g2-3a","Cyp6g2-3d"))
plot.data$metabolite <- factor(plot.data$metabolite,levels=c("Imidacloprid","5-OH","Olefin"))
plot.data$measurement_type <- factor(plot.data$measurement_type)
plot.data$measurement_value <- as.numeric(plot.data$measurement_value)
vars <- data.frame(expand.grid(levels(plot.data$metabolite), levels(plot.data$measurement_type)))
colnames(vars) <- c("metabolite", "measurement_type")
annot <- data.frame(x = rep(-Inf, dim(vars)[1]), y = rep(Inf, dim(vars)[1]), vars,labs=LETTERS[1:dim(vars)[1]])


equation.list <- vector("character",6)
for(m in unique(uas.cyp6g2.hplc.data.rearrange$metabolite)){
      for(t in (unique(uas.cyp6g2.hplc.data.rearrange$measurement_type))){
            sub <- subset(uas.cyp6g2.hplc.data.rearrange,metabolite==m & measurement_type==t)
            equation.list[intersect(which(annot$metabolite==m),which(annot$measurement_type==t))] <- equation.extract.hplc("measurement_value","genotype",data=sub,sub=F,bonf=6)
      }}
equation.list[intersect(which(annot$metabolite=="Imidacloprid"),which(annot$mes.type=="media"))] <- ""
annot <- mutate(annot,pval=equation.list)

```

Then it can be plotted in a similar way to Figure 3 and 5

```{r Plot Figure S5,fig.height=8,fig.width=5}


gp <- ggplot(plot.data,aes(x=genotype,y=measurement_value,fill=genotype))
gp <- gp+facet_grid(metabolite~measurement_type)
gp <- gp+ylim(c(0,2200000))
gp <- gp+ylab("Quantity of Chemical")
gp <- gp+stat_summary(geom="errorbar",colour='black',size=1,width=.4,fun.data="mean_se")
gp <- gp+stat_summary(fun.y="mean",fun.ymax="mean",fun.ymin="mean", geom="crossbar", width=.4,size=.3)
gp <- gp+geom_dotplot(binaxis='y', stackdir='center',dotsize=1.5) 
gp <- gp+geom_text(size=7,data=annot,aes(x, y, label=labs,fontface="bold",family="serif"),inherit.aes=FALSE,vjust=1.2,hjust=-.2)
gp <- gp+geom_text(size=7,data=annot,aes(1.5, y, label=pval,fontface="bold",family="serif"),inherit.aes=F,vjust=1.2)
gp <- gp+scale_fill_manual(values=c("black","orange4")) 
gp <- gp+theme_bw(base_size=14,base_family="serif")
gp <- gp+g1KO.wiggle_theme
print(gp)
```

